### PR TITLE
Feature1- Create a Database in Azure Cosmos DB 

### PR DESCRIPTION
**Pull request by sachinira**
This contains the creation of new database in Cosmos DB resource. It uses the SQL API REST endpoint.

### CODE REVIEW

File: src/azure_cosmosdb/cosmosdb_data_mappings.bal [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+isolated function mapParametersToHeaderType(string httpVerb, string url) returns RequestHeaderParameters {
+    RequestHeaderParameters params = {};
+    params.verb = httpVerb;
+    params.resourceType = getResourceType(url);
+    params.resourceId = getResourceId(url);
+    return params;
+}
+
+isolated function mapResponseHeadersToObject(http:Response|http:ClientError httpResponse) returns @tainted Headers|error{
```
Comment (LakshanSS) [inline src/azure_cosmosdb/cosmosdb_data_mappings.bal:11]: Shall we add a space before the open curly bracket?
Please check this with all the functions.

File: src/azure_cosmosdb/cosmosdb_data_mappings.bal [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    if (httpResponse is http:Response) {
+        responseHeaders.continuationHeader = getHeaderIfExist(httpResponse,"x-ms-continuation");
+        responseHeaders.sessionTokenHeader = getHeaderIfExist(httpResponse,"x-ms-session-token");
+        responseHeaders.requestChargeHeader = getHeaderIfExist(httpResponse,"x-ms-request-charge");
+        responseHeaders.resourceUsageHeader = getHeaderIfExist(httpResponse,"x-ms-resource-usage");
+        responseHeaders.itemCountHeader = getHeaderIfExist(httpResponse,"x-ms-item-count");
+        responseHeaders.etagHeader = getHeaderIfExist(httpResponse,"etag");
+        responseHeaders.dateHeader = getHeaderIfExist(httpResponse,"Date");
+        return responseHeaders;
+
```
Comment (LakshanSS) [inline src/azure_cosmosdb/cosmosdb_data_mappings.bal:22]: Shall we remove this unwanted newline?

File: src/azure_cosmosdb/cosmosdb_data_mappings.bal [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+isolated function mapJsonToDatabaseType([json, Headers] jsonPayload) returns Database {
+    json payload;
+    Headers headers;
+    [payload,headers] = jsonPayload;
+    Database db = {};
+    db.id = payload.id.toString();
+    db.reponseHeaders = headers;
+    return db;
+}
```
Comment (LakshanSS) [inline src/azure_cosmosdb/cosmosdb_data_mappings.bal:36]: Add a newline in the end of the file. 
In any file, we have to add one newline at the end of the file.

File: src/azure_cosmosdb/cosmosdb_endpoint.bal [hunk 4]
```diff
@@ -0,0 +1,48 @@
+import ballerina/http;
+
+# Azure Cosmos DB Client object.
+# + azureCosmosClient - The HTTP Client
+public  client class Client {
+
```
Comment (LakshanSS) [inline src/azure_cosmosdb/cosmosdb_endpoint.bal:6]: Shall we remove this newline?

File: src/azure_cosmosdb/cosmosdb_endpoint.bal [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    
+    private string baseUrl;
+    private string masterKey;
+    private string host;
+    private string keyType;
+    private string tokenVersion;
+
+    public http:Client azureCosmosClient;
+
+    function init(AzureCosmosConfiguration azureConfig){
```
Comment (LakshanSS) [inline src/azure_cosmosdb/cosmosdb_endpoint.bal:15]: Space before open curly bracket. Check all the functions and fix this.

File: src/azure_cosmosdb/cosmosdb_utils.bal [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+import ballerina/lang.'string as str;
+
+# To construct resource type  which is used to create the hashed token signature 
+# + url - string parameter part of url to extract the resource type
+# + return - Returns the resource type extracted from url as a string  
+isolated function getResourceType(string url) returns string{
+    string resourceType = EMPTY_STRING;
+    string[] urlParts = stringutils:split(url,FORWARD_SLASH);
+    int count = urlParts.length()-1;
+    if count % 2 != 0{
```
Comment (LakshanSS) [inline src/azure_cosmosdb/cosmosdb_utils.bal:14]: space before open curly bracket :)

**Comment by kasthuriraajan** [comment]
Shall we remove the target directory? I think there should be a change in .gitignore file.

File: src/azure_cosmosdb/cosmosdb_data_mappings.bal [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+isolated function mapParametersToHeaderType(string httpVerb, string url) returns RequestHeaderParameters {
+    RequestHeaderParameters params = {};
+    params.verb = httpVerb;
+    params.resourceType = getResourceType(url);
+    params.resourceId = getResourceId(url);
+    return params;
+}
+
+isolated function mapResponseHeadersToObject(http:Response|http:ClientError httpResponse) returns @tainted Headers|error 
+{
```
Comment (LakshanSS) [inline src/azure_cosmosdb/cosmosdb_data_mappings.bal:12]: Shall we move this open curly bracket to the previous line? 
Check in all the places and fix similar issue

File: src/azure_cosmosdb/cosmosdb_utils.bal [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+isolated function setHeaders(http:Request req, string host, string keyToken, string tokenType, string tokenVersion,
+RequestHeaderParameters params) returns http:Request|error {
+    req.setHeader("x-ms-version",params.apiVersion);
+    req.setHeader("Host",host);
+    req.setHeader("Accept","*/*");
+    req.setHeader("Connection","keep-alive");
+
+    string?|error date = getTime();
+    if date is string
+    {
```
Comment (LakshanSS) [inline src/azure_cosmosdb/cosmosdb_utils.bal:91]: Shall we move the open bracket to the previous line?

File: src/azure_cosmosdb/cosmosdb_utils.bal [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+# To construct the hashed token signature for a token to set  'Authorization' header
+# + verb - HTTP verb, such as GET, POST, or PUT
+# + resourceType - identifies the type of resource that the request is for, Eg. "dbs", "colls", "docs"
+# + resourceId -dentity property of the resource that the request is directed at
+# + keyToken - master or resource token
+# + tokenType - denotes the type of token: master or resource.
+# + tokenVersion - denotes the version of the token, currently 1.0.
+# + return - If successful, returns string which is the  hashed token signature. Else returns ().  
+isolated function generateTokenNew(string verb, string resourceType, string resourceId, string keyToken, string tokenType, 
+string tokenVersion) returns string? {
```
Comment (LakshanSS) [inline src/azure_cosmosdb/cosmosdb_utils.bal:124]: Shall we format it properly?
(Can move the tokenVersion also to the previous line if id doesn't exceed 120 characters)

File: src/azure_cosmosdb/cosmosdb_utils.bal [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+    var token = generateTokenJava(java:fromString(verb),java:fromString(resourceType),java:fromString(resourceId),
+    java:fromString(keyToken),java:fromString(tokenType),java:fromString(tokenVersion));
+    return java:toString(token);
+
+}
+
+isolated function setThroughputOrAutopilotHeader(http:Request req, ThroughputProperties? throughputProperties) returns 
+http:Request|error {
```
Comment (LakshanSS) [inline src/azure_cosmosdb/cosmosdb_utils.bal:132]: Shall we move this to the previous line?

File: src/azure_cosmosdb/cosmosdb_utils.bal [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+            req.setHeader("x-ms-cosmos-offer-autopilot-settings",throughputProperties.maxThroughput.toString());
+        } else if throughputProperties.throughput is int &&  throughputProperties.maxThroughput != () {
+            return 
+            prepareError("Cannot set both x-ms-offer-throughput and x-ms-cosmos-offer-autopilot-settings headers at once");
+        }
+    }
+    return req;
+}
+
+
```
Comment (LakshanSS) [inline src/azure_cosmosdb/cosmosdb_utils.bal:148]: Can remove unwanted newline

File: src/azure_cosmosdb/cosmosdb_utils.bal [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+isolated function mapResponseToTuple(http:Response|http:ClientError httpResponse) returns  @tainted [json,Headers]|error
+{
```
Comment (LakshanSS) [inline src/azure_cosmosdb/cosmosdb_utils.bal:150]: Can move to the previous line

File: src/azure_cosmosdb/cosmosdb_utils.bal [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+        }
+    } else {
+        return prepareError("Error occurred while invoking the REST API");
+    }
+}
+
+isolated function getHeaderIfExist(http:Response httpResponse, string headername) returns @tainted string? {
+    if httpResponse.hasHeader(headername) {
+        return httpResponse.getHeader(headername);
+    }else {
```
Comment (LakshanSS) [inline src/azure_cosmosdb/cosmosdb_utils.bal:181]: can add a space before else

File: src/azure_cosmosdb/cosmosdb_utils.bal (hunk 11 above)
Comment (LakshanSS) [inline src/azure_cosmosdb/cosmosdb_utils.bal:148]: Can remove unwanted newline

File: src/azure_cosmosdb/cosmosdb_utils.bal [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+# + keyToken - master or resource token
+# + tokenType - denotes the type of token: master or resource.
+# + tokenVersion - denotes the version of the token, currently 1.0.
+# + return - If successful, returns string which is the  hashed token signature. Else returns ().  
+isolated function generateTokenNew(string verb, string resourceType, string resourceId, string keyToken, string tokenType, 
+string tokenVersion) returns string? {
+    var token = generateTokenJava(java:fromString(verb),java:fromString(resourceType),java:fromString(resourceId),
+    java:fromString(keyToken),java:fromString(tokenType),java:fromString(tokenVersion));
+    return java:toString(token);
+
```
Comment (LakshanSS) [inline src/azure_cosmosdb/cosmosdb_utils.bal:126]: Can remove this unwanted newline

File: src/azure_cosmosdb/cosmosdb_utils.bal [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+}
+
+isolated function setThroughputOrAutopilotHeader(http:Request req, ThroughputProperties? throughputProperties) returns 
+http:Request|error {
+
```
Comment (LakshanSS) [inline src/azure_cosmosdb/cosmosdb_utils.bal:131]: Can remove this unwanted newline

File: src/azure_cosmosdb/cosmosdb_utils.bal [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+        return httpResponse.getHeader(headername);
+    } else {
+        return ();
+    }
+}
+
+isolated function generateTokenJava(handle verb, handle resourceType, handle resourceId, handle keyToken, handle tokenType, 
+handle tokenVersion) returns handle = @java:Method {
+    name: "generate",
+    'class: "com.sachini.TokenCreate"
```
Comment (Chuhaa) [inline src/azure_cosmosdb/cosmosdb_utils.bal:183]: It is better to rename the package name.

**Comment by sachinira** [comment]
Closing the pull request to change the folder structure

**Pull request closed without merging**